### PR TITLE
Remove deprecated hiera and validation functions

### PR DESCRIPTION
**Pull request by jkroepke**
Related #118 

* Replace deprecated hiera functions with newer lookup functions
* Replace stdlib's validation functions with puppet data types (bump stdlib minimum version to 4.13)
* Removed unused fact `gitlab_systemd`
* Added `Dockerfile` to run rspec tests locally without setup a dev environment.

### CODE REVIEW

**Comment by jkroepke** [comment]
Need help here, I don't know why the yaml is generated wrong.

File: manifests/config.pp [hunk 1]
```diff
@@ -83,15 +83,15 @@
   if $config_manage {
     if $source_config_file {
       file { $config_file:
-        ensure  => file,
```
Comment (b4ldr) [inline manifests/config.pp]: why change this to present, file is much more descriptive and does the same thing (i thought puppet-lint flagged this but dont see it on their we site)

**Comment by b4ldr** [comment]
can you give more info on the issue, happy to help with this

File: manifests/config.pp (hunk 1 above)
Comment (jkroepke) [inline manifests/config.pp]: I don't know why I'm doing this the last six month. I will change it back.

**Comment by jkroepke** [comment]
I'll take a rebase and rollback the changes on the ensure attributes.

**Comment by jkroepke** [comment]
Sorry for some possible notification spam.

@tobru @bastelfreak 

Any chances to get merged this PR? Rebasing this PR everytime is a hard doing.

File: manifests/cirunner.pp [hunk 2]
```diff
@@ -82,12 +74,9 @@
...
         ensure_packages('apt-transport-https')
 
-        $distid = downcase($::lsbdistid)
-
-        ::apt::source { 'apt_gitlabci':
+        apt::source { 'apt_gitlabci':
           comment  => 'GitlabCI Runner Repo',
-          location => "${repo_base_url}/runner/${package_name}/${distid}/",
-          release  => $::lsbdistcodename,
+          location => "${repo_base_url}/runner/${package_name}/${::lsbdistid.downcase}/",
```
Comment (LongLiveCHIEF) [inline manifests/cirunner.pp]: This is a legacy fact. Can you switch it to a modern fact please?

File: manifests/custom_hook.pp [hunk 3]
```diff
@@ -58,36 +58,27 @@
...
   } elsif $::gitlab::git_data_dir {
     $_repos_path = "${::gitlab::git_data_dir}/repositories"
   } else {
     $_repos_path = '/var/opt/gitlab/git-data/repositories'
   }
-  validate_absolute_path($_repos_path)
 
   if ! ($content) and ! ($source) {
-    fail('gitlab::custom_hook resource must specify either content or source')
+    fail("gitlab::custom_hook[${name}]: Must specify either content or source")
```
Comment (LongLiveCHIEF) [inline manifests/custom_hook.pp]: Nice touch 👍

File: metadata.json [hunk 4]
```diff
@@ -68,7 +68,7 @@
   "requirements": [
     {
       "name": "puppet",
-      "version_requirement": ">=3.7.0 <5.0.0"
+      "version_requirement": ">=4.6.1 <5.0.0"
```
Comment (LongLiveCHIEF) [inline metadata.json]: shouldn't the `version_requirement` here now be `">=4.6.1 <6.0.0` instead of `5.0.0`?

File: manifests/cirunner.pp (hunk 2 above)
Comment (jkroepke) [reply manifests/cirunner.pp]: If you want to fix the sped tests, sure. I tried to used the modern facts, but the lib rspec-puppet-facts doesn‘t provide modern facts and I get tons of error in travis.

Comment (jkroepke) [reply manifests/cirunner.pp]: If you want to fix the sped tests, sure. I tried to used the modern facts, but the lib rspec-puppet-facts doesn‘t provide modern facts and I get tons of error in travis.

Comment (LongLiveCHIEF) [reply manifests/cirunner.pp]: nah, sounds like out of scope for this PR in that case. We'll fix it in another issue/PR.

File: metadata.json (hunk 4 above)
Comment (jkroepke) [reply metadata.json]: Yeah, I created this PR so long ago.

Comment (LongLiveCHIEF) [reply metadata.json]: I feel you. I've got a few of my own that have had to be rebased a handful of times.

**Comment by jkroepke** [comment]
So I fixed the lint issues from master, too.

**Pull request merged**